### PR TITLE
Add vendor-check script, Makefile target and CI step

### LINKED ISSUE #556: Create a vendor check for dependencies
# Problem

Currently, our CI process doesn't ensure that the dependencies in `vendor` folder are the same as in `Gopkg.lock` file.

# Implementation
The script should run `dep ensure` and then check if anything was changed in the `vendor` directory. If so — fail the pipeline.

The command to run this script should be `make vendor-check`

See [this thread](https://github.com/status-im/status-go/issues/223#issuecomment-357777118) for more practical advice.

1st party implementation is coming to `dep`: [golang/dep#121](https://github.com/golang/dep/issues/121).

# Acceptance Criteria

1. A makefile target is created that fails if there is something in a `vendor` folder that isn't in `Gopkg.lock`. Or if `dep ensure` changes something in the `vendor` folder.

2. `.travis.yml` is updated to have a step of checking dependencies before building or running any tests. (NOTE: Travis restarts builds on every push. If you need to force restarting it and have no changes to commit, just alter the latest commit message with `git commit --amend` and push. That should do the trick).

### PR DESCRIPTION
**Pull request by azer**
It adds `vendor-check` target, copying the validation script in [dep repository](https://github.com/golang/dep/blob/master/hack/validate-vendor.bash).

Closes #556 

CC @mandrigin

### CODE REVIEW

File: .validate-vendor.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		{
+			echo 'The contents of vendor differ after "dep ensure":'
+			echo
+			echo "$diffs"
+			echo
+			echo 'Make sure these commands have been run before committing.'
+			echo
+		} >&2
+		false
+  else
```
Comment (mandrigin) [inline .validate-vendor.sh]: Seems like indentation went funky in this file, let's fix that.

File: .validate-vendor.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+# for our repo and makes sure that it was done in a valid way.
+#
+# This file is a slightly modified copy of https://github.com/golang/dep/blob/master/hack/validate-vendor.bash
+# The only change made was checking if `dep` was available and installing it if necessary.
+
+
+set -e -o pipefail
+
+if [ -z "$VALIDATE_UPSTREAM" ]; then
+	VALIDATE_REPO='https://github.com/golang/dep.git'
```
Comment (mandrigin) [inline .validate-vendor.sh]: It should be the `status-go` repo, right?

File: .validate-vendor.sh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	VALIDATE_BRANCH='master'
```
Comment (mandrigin) [inline .validate-vendor.sh]: We use `develop` as a target branch.

**Review by mandrigin — CHANGES_REQUESTED** [review]
Looks good! A few changes are necessary.

Also, it would be nice to validate that the script actually returns a non-zero answer if contents of `vendor` are changed. Now it shows `warning: no common commits` because it checks against the wrong repo.

**Comment by mandrigin** [comment]
@azer looks good now!
One last point. It fails on Travis now due to `dep` not being installed. Let's add `dep-install` as a Makefile target and add it to the `install:` phase of `.travis.yml`.

**Comment by azer** [comment]
@mandrigin Thanks for the feedback, I've made the fixes requested, also tested it on [a branch with additions into `vendor/` folder](https://github.com/azer/status-go/tree/vendor-changed/). Here is what the failure exit looks like;

```
The contents of vendor differ after "dep ensure":

 D vendor/github.com/azer/logger/README.md
 D vendor/github.com/azer/logger/examples/attrs.go
 D vendor/github.com/azer/logger/examples/multiple.go
 D vendor/github.com/azer/logger/examples/programmatical.go
 D vendor/github.com/azer/logger/examples/simple.go
 D vendor/github.com/azer/logger/examples/timer-performance.go
 D vendor/github.com/azer/logger/format.go
 D vendor/github.com/azer/logger/log.go
 D vendor/github.com/azer/logger/logger.go
 D vendor/github.com/azer/logger/runtime.go
 D vendor/github.com/azer/logger/standard-output.go

Make sure these commands have been run before committing.
```

**Review by mandrigin — APPROVED** [review]

**Comment by mandrigin** [comment]
@azer Looks good! I'll ask one peer of mine to review it too.

File: ci/validate-vendor.sh [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env bash
+# Copyright 2017 The Go Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file.
+#
+# This script checks if we changed anything with regard to dependency management
+# for our repo and makes sure that it was done in a valid way.
+#
+# This file is a slightly modified copy of https://github.com/golang/dep/blob/master/hack/validate-vendor.bash
+# The only change made was checking if `dep` was available and installing it if necessary.
```
Comment (themue) [inline ci/validate-vendor.sh]: Isn't this check done with `dep-install` in `Makefile`?

**Review by themue — APPROVED** [review]
Only one question, rest LGTM.

File: ci/validate-vendor.sh [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+
+set -e -o pipefail
+
+if [ -z "$VALIDATE_UPSTREAM" ]; then
+	VALIDATE_REPO='https://github.com/status-im/status-go'
+	VALIDATE_BRANCH='develop'
```
Comment (adambabik) [inline ci/validate-vendor.sh]: Shouldn't branch be taken from env and only default to `develop`? We merge almost every branch to `develop` but it should change soon.

File: ci/validate-vendor.sh [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+	validate_diff() {
+		if [ "$VALIDATE_UPSTREAM" != "$VALIDATE_HEAD" ]; then
+		    git diff "$VALIDATE_COMMIT_DIFF" "$@"
+		fi
+	}
+fi
+
+IFS=$'\n'
+files=( $(validate_diff --diff-filter=ACMR --name-only -- 'Gopkg.toml' 'Gopkg.lock' 'vendor/' || true) )
```
Comment (adambabik) [inline ci/validate-vendor.sh:37]: If the `if` statement above won't be executed, `validate_diff` won't be defined, right?

File: ci/validate-vendor.sh [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+# This script checks if we changed anything with regard to dependency management
+# for our repo and makes sure that it was done in a valid way.
+#
+# This file is a slightly modified copy of https://github.com/golang/dep/blob/master/hack/validate-vendor.bash
+# The only change made was checking if `dep` was available and installing it if necessary.
+
+
+set -e -o pipefail
+
+if [ -z "$VALIDATE_UPSTREAM" ]; then
```
Comment (adambabik) [inline ci/validate-vendor.sh]: I hate bash for their super hard to remember expression. Can you please add a comment `// if $VALIDATE_UPSTREAM is null or empty string do bla bla bla...`?

File: ci/validate-vendor.sh [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+		    git diff "$VALIDATE_COMMIT_DIFF" "$@"
+		fi
+	}
+fi
+
+IFS=$'\n'
+files=( $(validate_diff --diff-filter=ACMR --name-only -- 'Gopkg.toml' 'Gopkg.lock' 'vendor/' || true) )
+unset IFS
+
+if [ ${#files[@]} -gt 0 ]; then
```
Comment (adambabik) [inline ci/validate-vendor.sh]: I have no idea what `${#files[@]}` does (getting a number of files?). Can you please add a comment?

File: ci/validate-vendor.sh [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+		{
+			echo 'The contents of vendor differ after "dep ensure":'
+			echo
+			echo "$diffs"
+			echo
+			echo 'Make sure these commands have been run before committing.'
+			echo
+		} >&2
+		false
+  else
```
Comment (adambabik) [inline ci/validate-vendor.sh]: Formatting? Shouldn't this `else` be aligned with `if [ "$diffs" ]; then`?

**Review by adambabik — CHANGES_REQUESTED** [review]

**Comment by adambabik** [comment]
I know it's copied from `dep` repo but it would be great to add a general comment what this bash script actually does in 2-3 sentences.

File: ci/validate-vendor.sh (hunk 4 above)
Comment (azer) [reply ci/validate-vendor.sh]: Yup, just deleted it

File: ci/validate-vendor.sh (hunk 7 above)
Comment (azer) [reply ci/validate-vendor.sh]: Yes, I've added some more comments

File: ci/validate-vendor.sh (hunk 6 above)
Comment (azer) [reply ci/validate-vendor.sh:37]: Yes, if statement will always be executed unless we define `VALIDATE_UPSTREAM` before calling this script.

File: ci/validate-vendor.sh (hunk 8 above)
Comment (azer) [reply ci/validate-vendor.sh]: Splits the content of `files` variable by whitespaces and gets the number of the lines. Just added a comment for it.

File: ci/validate-vendor.sh (hunk 9 above)
Comment (azer) [reply ci/validate-vendor.sh]: Fixed 👍

**Comment by azer** [comment]
@adambabik defaulted branch to "develop", added some comments & fixed tabbing 👍

File: ci/validate-vendor.sh (hunk 5 above)
Comment (azer) [reply ci/validate-vendor.sh]: 👍

File: ci/validate-vendor.sh (hunk 7 above)
Comment (mandrigin) [reply ci/validate-vendor.sh]: Hmm, do we even need this option?

**Review by adambabik — APPROVED** [review]
Thanks @azer. LGTM

**Comment by mandrigin** [comment]
@azer thanks for your contribution 🎉

**Pull request merged**